README.rst
============
Introduction
============

.. image:: https://img.shields.io/travis/nazrulworld/fhirpath.svg
        :target: https://travis-ci.org/nazrulworld/fhirpath

.. image:: https://readthedocs.org/projects/fhirpath/badge/?version=latest
        :target: https://fhirpath.readthedocs.io/en/latest/?badge=latest
        :alt: Documentation Status

.. image:: https://codecov.io/gh/nazrulworld/fhirpath/branch/master/graph/badge.svg
   :target: https://codecov.io/gh/nazrulworld/fhirpath/branch/master
   :alt: Test Coverage

.. image:: https://img.shields.io/pypi/pyversions/fhirpath.svg
   :target: https://pypi.python.org/pypi/fhirpath/
   :alt: Python Versions

.. image:: https://img.shields.io/lgtm/grade/python/g/nazrulworld/fhirpath.svg?logo=lgtm&logoWidth=18
    :target: https://lgtm.com/projects/g/nazrulworld/fhirpath/context:python
    :alt: Language grade: Python

.. image:: https://img.shields.io/pypi/v/fhirpath.svg
   :target: https://pypi.python.org/pypi/fhirpath

.. image:: https://img.shields.io/pypi/l/fhirpath.svg
   :target: https://pypi.python.org/pypi/fhirpath/
   :alt: License

.. image:: https://static.pepy.tech/personalized-badge/fhirpath?period=total&units=international_system&left_color=black&right_color=green&left_text=Downloads
    :target: https://pepy.tech/project/fhirpath
    :alt: Downloads

.. image:: https://www.hl7.org/fhir/assets/images/fhir-logo-www.png
        :target: https://www.hl7.org/fhir/fhirpath.html
        :alt: HL7® FHIR®

FHIRPath_ Normative Release (v2.0.0) implementation in Python, along side it
provides support for `FHIR Search <https://www.hl7.org/fhir/search.html>`_ API and
Query (we called it ``fql(FHIR Query Language)``)
API to fetch FHIR resources from any data-source(database).
This library is built in ORM_ like approach. Our goal is to make 100% (as much as possible)
FHIRPath_ Normative Release (v2.0.0) specification compliance product.

* Supports FHIR® ``STU3`` and ``R4``.
* Supports multiple provider´s engine. Now Plone_ & guillotina_ framework powered providers `fhirpath-guillotina`_ and `collective.fhirpath`_ respectively are supported and more coming soon.
* Supports multiple dialects, for example elasticsearch_, GraphQL_, PostgreSQL_. Although now elasticsearch_ has been supported.
* Provide full support of `FHIR Search <https://www.hl7.org/fhir/search.html>`_ with easy to use API.


Usages
------

This library is kind of abstract type, where all specifications from FHIRPath_ Normative Release (v2.0.0) are implemented rather than completed solution (ready to go).
The main reason behind this design pattern, to support multiple database systems as well as well as any framework, there is no dependency.

``fhirpath`` never taking care of creating indexes, mappings (elasticsearch) and storing data, if you want to use this library, you have to go
through any of existing providers (see list bellow) or make your own provider (should not too hard work).


Simple example
~~~~~~~~~~~~~~

Assumption:

1. Elasticsearch server 7.x.x Installed.

2. Mappings and indexes are handled manually.

3. Data (document) also are stored manually.


Create Connection and Engine::

    >>> from fhirpath.connectors import create_connection
    >>> from fhirpath.engine.es import ElasticsearchEngine
    >>> from fhirpath.engine import dialect_factory
    >>> from fhirpath.enums import FHIR_VERSION

    >>> host, port = "127.0.0.1", 9200
    >>> conn_str = "es://@{0}:{1}/".format(host, port)
    >>> connection = create_connection(conn_str, "elasticsearch.Elasticsearch")
    >>> connection.raw_connection.ping()
    True
    >>> engine = ElasticsearchEngine(FHIR_VERSION.R4, lambda x: connection, dialect_factory)


Basic Search::

    >>> from fhirpath.search import Search
    >>> from fhirpath.search import SearchContext

    >>> search_context = SearchContext(engine, "Organization")
    >>> params = (
    ....    ("active", "true"),
    ....    ("_lastUpdated", "2010-05-28T05:35:56+00:00"),
    ....    ("_profile", "http://hl7.org/fhir/Organization"),
    ....    ("identifier", "urn:oid:2.16.528.1|91654"),
    ....    ("type", "http://hl7.org/fhir/organization-type|prov"),
    ....    ("address-postalcode", "9100 AA"),
    ....    ("address", "Den Burg"),
    .... )
    >>> fhir_search = Search(search_context, params=params)
    >>> bundle = fhir_search()
    >>> len(bundle.entry) == 0
    True

Basic Query::

    >>> from fhirpath.enums import SortOrderType
    >>> from fhirpath.query import Q_
    >>> from fhirpath.fql import T_
    >>> from fhirpath.fql import V_
    >>> from fhirpath.fql import exists_
    >>> query_builder = Q_(resource="Organization", engine=engine)
    >>>  query_builder = (
    ....    query_builder.where(T_("Organization.active") == V_("true"))
    ....    .where(T_("Organization.meta.lastUpdated", "2010-05-28T05:35:56+00:00"))
    ....    .sort(sort_("Organization.meta.lastUpdated", SortOrderType.DESC))
    .... )
    >>> query_result = query_builder(async_result=False)
    >>> for resource in query_result:
    ....    assert resource.__class__.__name__ == "OrganizationModel"
    >>> # test fetch all
    >>> result = query_result.fetchall()
    >>> result.__class__.__name__ == "EngineResult"
    True

    >>> query_builder = Q_(resource="ChargeItem", engine=engine)
    >>> query_builder = query_builder.where(exists_("ChargeItem.enteredDate"))
    >>> result = query_builder(async_result=False).single()
    >>> result is not None
    True
    >>> isinstance(result, builder._from[0][1])
    True

    >>> query_builder = Q_(resource="ChargeItem", engine=engine)
    >>> query_builder = query_builder.where(exists_("ChargeItem.enteredDate"))
    >>> result = query_builder(async_result=False).first()
    >>> result is not None
    True
    >>> isinstance(result, builder._from[0][1])
    True


Available Provider (known)
--------------------------

Currently very few numbers of providers available, however more will coming soon.

`fhirpath-guillotina`_
~~~~~~~~~~~~~~~~~~~~~~

A `guillotina`_ framework powered provider, battery included, ready to go! `Please follow associated documentation. <https://fhirpath-guillotina.readthedocs.io/en/latest/>`_

1. **Engine**: Elasticsearch

2. **PyPi**: https://pypi.org/project/fhirpath-guillotina/

3. **Source**: https://github.com/nazrulworld/fhirpath_guillotina


`collective.fhirpath`_
~~~~~~~~~~~~~~~~~~~~~~

A `Plone`_ powered provider, like `fhirpath-guillotina`_ every thing is included. ready to go, although has a dependency
on `plone.app.fhirfield`_.

1. **Engine**: Elasticsearch

2. **PyPi**: https://pypi.org/project/collective.fhirpath/

3. **Source**: https://github.com/nazrulworld/collective.fhirpath


unlisted
~~~~~~~~
Why are you waiting for? You are welcome to list your provider here!
Developing provider should not be so hard, as ``fhirpath`` is giving you convenient APIs.


Elasticsearch Custom Analyzer
-----------------------------
To get some special search features for reference type field, you will need to setup custom analyzer for your elasticsearch index.

Example Custom Analyzer::

    settings = {
        "analysis": {
            "normalizer": {
                "fhir_token_normalizer": {"filter": ["lowercase", "asciifolding"]}
            },
            "analyzer": {
                "fhir_reference_analyzer": {
                    "tokenizer": "keyword",
                    "filter": ["fhir_reference_filter"],
                },
            },
            "filter": {
                "fhir_reference_filter": {
                    "type": "pattern_capture",
                    "preserve_original": True,
                    "patterns": [r"(?:\w+\/)?(https?\:\/\/.*|[a-zA-Z0-9_-]+)"],
                },
            },
            "char_filter": {},
            "tokenizer": {},
        }


Example Mapping (Reference Field)::

    "properties": {
      "reference": {
        "type": "text",
        "index": true,
        "store": false,
        "analyzer": "fhir_reference_analyzer"
    }


ToDo
----

1. `fhirbase`_ engine aka provider implementation.

2. All methods/functions are defined in `FHIRPath`_ specification, would be completed.

3. Implement https://github.com/ijl/orjson
4. https://developers.redhat.com/blog/2017/11/16/speed-python-using-rust/

Credits
-------

This package skeleton was created with Cookiecutter_ and the `audreyr/cookiecutter-pypackage`_ project template.

.. _Cookiecutter: https://github.com/audreyr/cookiecutter
.. _`audreyr/cookiecutter-pypackage`: https://github.com/audreyr/cookiecutter-pypackage
.. _`FHIRPath`: http://hl7.org/fhirpath/N1/
.. _`FHIR`: http://hl7.org/fhir/
.. _`ORM`: https://en.wikipedia.org/wiki/Object-relational_mapping
.. _`Plone`: https://plone.org
.. _`guillotina`: https://guillotina.readthedocs.io/en/latest/
.. _`elasticsearch`: https://www.elastic.co/products/elasticsearch
.. _`GraphQL`: https://graphql.org/
.. _`PostgreSQL`: https://www.postgresql.org/
.. _`fhirpath-guillotina`: https://pypi.org/project/fhirpath-guillotina/
.. _`collective.fhirpath`: https://pypi.org/project/collective.fhirpath/
.. _`plone.app.fhirfield`: https://pypi.org/project/plone.app.fhirfield/
.. _`fhirbase`: https://github.com/fhirbase/fhirbase


© Copyright HL7® logo, FHIR® logo and the flaming fire are registered trademarks
owned by `Health Level Seven International <https://www.hl7.org/legal/trademarks.cfm?ref=https://pypi.org/project/fhir-resources/>`_

**"FHIR® is the registered trademark of HL7 and is used with the permission of HL7.
Use of the FHIR trademark does not constitute endorsement of this product by HL7"**


HISTORY.rst
=======
History
=======

0.10.6 (unreleased)
-------------------

- Switch to Apache Software License v2


0.10.5 (2020-12-17)
-------------------

Improvement

- ``BundleWrapper`` is providing two optional methods, ``calculate_fhir_base_url`` and ``resolve_absolute_uri`` and is now also taking optional parameter ``base_url``.


Fixes
- fixes ``ElasticSearchDialect.create_term`` [Kartik Sayani]

- fixes ``EngineResult.extract_references``. [Jason Paumier]

- fixes how composite search params are parsed. [Jason Paumier]

- Issue#28 `Nested GroupTerm search returns no matches <https://github.com/nazrulworld/fhirpath/issues/28>`_

- fixes ``SearchContext.normalize_param`` for composite search params [nazrulworld]

0.10.4 (2020-11-19)
-------------------

- fixes ``FHIRAbstractModel`` comparing at ``utils`` module for BundleWrapper.

- ``fallback_callable`` helper function is available at ``utils`` module.


0.10.3 (2020-11-17)
-------------------

Improvements

- More helper functions (``get_local_timezone``, ``timestamp_utc``, ``timestamp_utc``) are created.

- ``initial_bundle_data`` method is now available in Base Elasticsearch engine class,
  meaning that it is possible construct Bundle initial data into the derived class, so more flexibility.

Bugfixes

- Default bundle initial data was constructed ``meta.lastUpdated`` value with utc now time but without timezone offset, so
  during json serialization, timezone info was missed as a result reverse construct of Bundle complains validation error.

0.10.2 (2020-11-06)
-------------------

Improvements

- ``orjson`` is no longer required. ``json_dumps`` and ``json_loads`` now dynamically supports
  orjson and simplejson.


0.10.1 (2020-11-04)
-------------------

Bugfixes

- ``Connection.raw_connection`` was wrongly wrapped by ``AsyncElasticsearchConnection/ElasticsearchConnection.from_url()`` with self, instead of ``elasticsearch.AsyncElasticsearch/elasticsearch.Elasticsearch``'s instance.


0.10.0 (2020-11-04)
-------------------

Improvements


- Introducing `AsyncElasticsearchConnection`` and ``AsyncElasticsearchEngine`` the asyncio based connection and engine for Elasticsearch. See `Using Asyncio with Elasticsearch <https://elasticsearch-py.readthedocs.io/en/7.9.1/async.html>`_

- Added ``orjson`` based json serializer for Elasticsearch by default (when connection is made from connection factory).

- Added support for `_summary=text|data|count|true|false`. [arkhn]

- Added support for `_elements` search parameter. [arkhn]


Breaking

- ``async_result`` parameter is no longer needed for SearchContext, Search and Query (included async version) as from now all
  engine contains that information (``engine_class.is_async()``).

0.9.1 (2020-10-24)
------------------

- Added supports for ``_format`` and ``_pretty`` params, now there should no complain for those, instead of simply ignored. [nazrulworld]


0.9.0 (2020-10-24)
------------------

- Handle ``:identifier`` modifier for reference search parameters [simonvadee]

- fixes `BundleWrapper`` as_json mode, now includes with ``resourceType`` value. [nazrulworld]

- ``Dict`` response option has bee added in ``fhirpath.search.fhir_search`` [nazrulworld]

- Ignore empty search params #21 [simonvadee]

- Just for performance optimization issue minimum required ``zope.interface`` version is ``5.1.2``.

0.8.1 (2020-10-05)
------------------

- Disable pydantic validation for Bundle in fhirpath.utils.BundleWrapper [simonvadee]

- Two helper functions ``json_dumps`` and ``json_loads`` are now available under utils module [nazrulworld]

- Only apply search prefix on affected types #17 [simonvadee]

0.8.0 (2020-09-25)
------------------

Improvements

- add supports for some important FHIR search parameters (``_has``, ``_include`` and ``_revinclude``) [simonvadee]

- enable search on several resource types (_type search param) [Jasopaum]

- Issue #8 `Add search support for without any params or query string if context has resource type <https://github.com/nazrulworld/fhirpath/issues/8>`_ [nazrulworld]

- Issue #9 `multiple negative not working <https://github.com/nazrulworld/fhirpath/issues/9>`_ [nazrulworld]

Breaking

- ``fhirpath.search.SearchContext.resource_name`` has been changed ``fhirpath.search.SearchContext.resource_type`` and
  now datatype is List instead of string. Please check your API. [Jasopaum]

- For case of ``Elasticsearch`` based engine, you should use custom analyzer (``fhir_reference_analyzer``) for FHIR Reference type. For details see readme.


0.7.1 (2020-08-07)
------------------

- added missing ``isodate`` package dependency.


0.7.0 (2020-08-07)
------------------

Improvements

- Issue#5: Now ``ElasticsearchEngine::get_index_name`` takes one optional parameter ``resource_type``.

- Add supports for python version 3.6.

Breaking

- Make full capability with `fhir.resources <https://pypi.org/project/fhir.resources/>`_ version ``6.x.x``,
  please have a look of revolutionary changes of ``fhir.resources``.

0.6.2 (2020-06-30)
------------------

- ``fhirspec`` and ``fhir.resources`` versions are pinned.


0.6.1 (2020-05-09)
------------------
A must update release (from ``0.6.0``)!

Bugfixes

- fix: issues, those arieses due to fix bellow issue.

src/fhirpath/types.py
# _*_ coding: utf-8 _*_
"""©FHIR Data Primitive Types
https://www.hl7.org/fhir/datatypes.html#primitive
"""
import base64
import re
from collections import deque
from datetime import date, datetime
from typing import Deque, Optional, Text, Union

import isodate
from zope.interface import implementer

from fhirpath.thirdparty import ImmutableDict

from .interfaces import IFhirPrimitiveType, IPrimitiveTypeCollection, ITypeSpecifier

__author__ = "Md Nazrul Islam <email2nazrul@gmail.com>"

__all__ = [
    "EMPTY_VALUE",
    "FhirBoolean",
    "FhirInteger",
    "FhirString",
    "FhirDecimal",
    "FhirURI",
    "FhirURL",
    "FhirCanonical",
    "FhirBase64Binary",
    "FhirInstant",
    "FhirDate",
    "FhirDateTime",
    "FhirTime",
    "FhirCode",
    "FhirOid",
    "FhirId",
    "FhirMarkdown",
    "FhirUnsignedInt",
    "FhirPositiveInt",
    "FhirUUID",
    "PrimitiveDataTypes",
]


@implementer(IFhirPrimitiveType)
class FhirPrimitiveType(str):
    """In FHIR, the data types are divided into ‘primitive’ and ‘complex’ data types.
    The primitive data types are types like string, integer, boolean, etc.
    that can take a single value. The complex types consist of multiple values
    grouped together.
    For each of the fields that take a primitive data type, the API provides you with
    two fields in the class. One of the fields has the same name as the element it
    corresponds with in the FHIR resource,
    e.g. Active in the Patient class. This field is of the standard .Net data type."""

    __visit_name__: Text
    __regex__: Optional[Text]

    def _validate(self) -> None:
        """ """
        if self.__regex__ is not None:
            res = re.match(self.__regex__, self)
            if not res:
                raise ValueError(
                    "Invalid FHIR '{0}' value!".format(self.__visit_name__)
                )


class FhirBoolean(FhirPrimitiveType):
    """XML Representation: xs:boolean, except that 0 and 1 are not valid values
    JSON representation: JSON boolean (true or false)
    """

    __visit_name__: str = "boolean"
    __regex__: str = r"true|false"

    def _validate(self) -> None:
        """ """
        res = re.match(self.__regex__, self)
        if not res:
            raise ValueError(
                f"Invalid boolean value: expected true or false, got {self}"
            )

    def to_python(self) -> bool:
        """ """
        self._validate()

        return str(self) == "true"


class FhirInteger(FhirPrimitiveType):
    """A signed integer in the range −2,147,483,648..2,147,483,647
    (32-bit; for larger values, use decimal)

    XML Representation: xs:int, except that leading 0 digits are not allowed
    JSON representation: JSON number (with no decimal point)
    """

    __visit_name__: str = "integer"
    __regex__: str = r"[0]|[-+]?[1-9][0-9]*"

    def _validate(self) -> None:
        """ """
        res = re.match(self.__regex__, self)
        if not res:
            raise ValueError(
                "Invalid FHIR integer value! A signed integer in the"
                "range −2,147,483,648..2,147,483,647"
            )

    def to_python(self) -> int:
        """ """
        self._validate()

        return int(self)


class FhirString(FhirPrimitiveType):
    """A sequence of Unicode characters
    Note that strings SHALL NOT exceed 1MB (1024*1024 characters) in size.
    Strings SHOULD not contain Unicode character points below 32, except for u0009
    (horizontal tab), u0010 (carriage return) and u0013 (line feed).
    Leading and Trailing whitespace is allowed,
    but SHOULD be removed when using the XML format.
    Note: This means that a string that consists only of whitespace
    could be trimmed to nothing, which would be treated as an invalid element value.
    Therefore strings SHOULD always contain non-whitespace content

    XML Representation: xs:string
    JSON representation: JSON String
    """

    __visit_name__: str = "string"
    __regex__: str = r"[ \r\n\t\S]+"

    def _validate(self) -> None:
        """ """
        res = re.match(self.__regex__, self)
        if not res:
            raise ValueError(
                "Invalid FHIR integer value! A signed integer in the"
                "range −2,147,483,648..2,147,483,647"
            )

    def to_python(self) -> str:
        """ """
        return str(self)


class FhirDecimal(FhirPrimitiveType):
    """Rational numbers that have a decimal representation.
    The precision of the decimal value has significance:

    - e.g. 0.010 is regarded as different to 0.01, and the original precision
      should be preserved
    - Implementations SHALL handle decimal values in ways that preserve and respect
      the precision of the value as represented for presentation purposes
    - Implementations are not required to perform calculations with these
      numbers differently, though they may choose to do so (i.e. preserve significance)
    - In object code, implementations that might meet this constraint are
      GMP implementations or equivalents to Java BigDecimal that implement arbitrary
      precision, or a combination of a (64 bit) floating point value with a
      precision field
    - Note that large and/or highly precise values are extremely rare in medicine.
      One element where highly precise decimals may be encountered is the Location
      coordinates. Irrespective of this, the limits documented in XML Schema apply

    XML Representation: union of xs:decimal and xs:double (see below for limitations)
    JSON representation: A JSON number (see below for limitations)
    """

    __visit_name__: str = "decimal"
    __regex__: str = r"-?(0|[1-9][0-9]*)(\.[0-9]+)?([eE][+-]?[0-9]+)?"

    def _validate(self):
        """ """
        res = re.match(self.__regex__, self)
        if not res:
            raise ValueError("Invalid FHIR decimal value!")

    def to_python(self) -> float:
        """ """
        self._validate()

        return float(self)


class FhirURI(FhirPrimitiveType):
    """A Uniform Resource Identifier Reference (RFC 3986 ).
    Note: URIs are case sensitive.
    For UUID (urn:uuid:53fefa32-fcbb-4ff8-8a92-55ee120877b7) use all lowercase

    XML Representation: xs:anyURI
    JSON representation: A JSON string - a URI
    """

    __visit_name__: str = "uri"
    __regex__: str = r"\S*"

    def to_python(self) -> str:
        """"""
        return str(self)


class FhirURL(FhirPrimitiveType):
    """A Uniform Resource Locator (RFC 1738 ).
    Note URLs are accessed directly using the specified protocol.
    Common URL protocols are ``http{s}:``, ``ftp:``, ``mailto``: and ``mllp:``,
    though many others are defined

    XML Representation: xs:anyURI
    JSON representation: A JSON string - a URL
    """

    __visit_name__: str = "url"
    # xxx: restricted to defined protocol
    __regex__: Optional[str] = None

    def to_python(self) -> str:
        """ """
        return str(self)


class FhirCanonical(FhirPrimitiveType):
    """A URI that refers to a resource by its canonical URL (resources with a url property).
    The canonical type differs from a uri in that it has special meaning in this
    specification, and in that it may have a version appended, separated by
    avertical bar (|).Note that the type canonical is not used for the actual canonical
    URLs that are the target of these references, but for the URIs that refer to them,
    and may have the version suffix in them. Like other URIs, elements of type
    canonical may also have #fragment references

    XML Representation: xs:anyURI
    JSON representation: A JSON string - a canonical URL
    """

    __visit_name__: str = "canonical"

    __regex__: Optional[str] = None

    def to_python(self) -> Text:
        """ """
        return str(self)


@implementer(IFhirPrimitiveType)
class FhirBase64Binary(bytes):
    """A stream of bytes, base64 encoded (RFC 4648 )
    There is no specified upper limit to the size of a binary,
    but systems will have to impose some implementation based limit
    to the size they support. This should be clearly documented, though
    there is no computable for this at this time

    XML Representation: xs:base64Binary
    JSON representation: A JSON string - base64 content
    """

    __visit_name__: str = "base64Binary"
    __regex__: str = r"(\s*([0-9a-zA-Z\+\=]){4}\s*)+"

    def _validate(self) -> None:
        """ """
        res = re.match(self.__regex__, self.decode())
        if not res:
            raise ValueError("Invalid FHIR base64Binary value!")

    def to_python(self) -> bytes:
        """ """
        self._validate()

        return base64.b64decode(self)


class FhirInstant(FhirPrimitiveType):
    """An instant in time in the format YYYY-MM-DDThh:mm:ss.sss+zz:zz
    (e.g. 2015-02-07T13:28:17.239+02:00 or 2017-01-01T00:00:00Z).
    The time SHALL specified at least to the second and SHALL include a time zone.
    Note: This is intended for when precisely observed times are required (typically
    system logs etc.), and not human-reported times -
    for those, use date or dateTime (which can be as precise as instant,
    but is not required to be). instant is a more constrained dateTime

    XML Representation: xs:dateTime
    JSON representation: A JSON string - an xs:dateTime
    """

    __visit_name__: str = "instant"
    __regex__: str = r"([0-9]([0-9]([0-9][1-9]|[1-9]0)|[1-9]00)|[1-9]000)-(0[1-9]|1[0-2])-(0[1-9]|[1-2][0-9]|3[0-1])T([01][0-9]|2[0-3]):[0-5][0-9]:([0-5][0-9]|60)(\.[0-9]+)?(Z|(\+|-)((0[0-9]|1[0-3]):[0-5][0-9]|14:00))"  # noqa: E501

    def to_python(self) -> datetime:
        """ """
        self._validate()

        if "T" in self:
            return isodate.parse_datetime(self)
        else:
            return isodate.parse_date(self)


class FhirDate(FhirPrimitiveType):
    """A date, or partial date (e.g. just year or year + month) as
    used in human communication. The format is YYYY, YYYY-MM, or YYYY-MM-DD,
    e.g. 2018, 1973-06, or 1905-08-23. There SHALL be no time zone.
    Dates SHALL be valid dates

    XML Representation: union of xs:date, xs:gYearMonth, xs:gYear
    JSON representation: A JSON string - a union of xs:date, xs:gYearMonth, xs:gYear
    """

    __visit_name__: str = "date"
    __regex__: str = r"([0-9]([0-9]([0-9][1-9]|[1-9]0)|[1-9]00)|[1-9]000)(-(0[1-9]|1[0-2])(-(0[1-9]|[1-2][0-9]|3[0-1]))?)?"  # noqa: E501

    def to_python(self) -> date:
        """ """
        self._validate()

        return isodate.parse_date(self)


class FhirDateTime(FhirPrimitiveType):
    """A date, date-time or partial date (e.g. just year or year + month) as used in
    human communication. The format is YYYY, YYYY-MM, YYYY-MM-DD or
    YYYY-MM-DDThh:mm:ss+zz:zz, e.g. 2018, 1973-06, 1905-08-23,
    2015-02-07T13:28:17-05:00 or 2017-01-01T00:00:00.000Z.
    If hours and minutes are specified, a time zone SHALL be populated.
    Seconds must be provided due to schema type constraints but may be zero-filled
    and may be ignored at receiver discretion. Dates SHALL be valid dates.
    The time "24:00" is not allowed. Leap Seconds are allowed

    XML Representation: union of xs:dateTime, xs:date, xs:gYearMonth, xs:gYear
    JSON representation: A JSON string - a union of xs:dateTime,
    xs:date, xs:gYearMonth, xs:gYear
    """

    __visit_name__: str = "dateTime"
    __regex__: str = r"([0-9]([0-9]([0-9][1-9]|[1-9]0)|[1-9]00)|[1-9]000)(-(0[1-9]|1[0-2])(-(0[1-9]|[1-2][0-9]|3[0-1])(T([01][0-9]|2[0-3]):[0-5][0-9]:([0-5][0-9]|60)(\.[0-9]+)?(Z|(\+|-)((0[0-9]|1[0-3]):[0-5][0-9]|14:00)))?)?)?"  # noqa: E501

    def to_python(self) -> datetime:
        """ """
        self._validate()

        return isodate.parse_datetime(self)


class FhirTime(FhirPrimitiveType):
    """A time during the day, in the format hh:mm:ss.
    There is no date specified.
    Seconds must be provided due to schema type constraints but may be zero-filled and
    may be ignored at receiver discretion. The time "24:00" SHALL NOT be used.
    A time zone SHALL NOT be present.
    Times can be converted to a Duration since midnight.

    XML Representation: xs:time
    JSON representation: A JSON string - an xs:time
    """

    __visit_name__: str = "time"
    __regex__: str = r"([01][0-9]|2[0-3]):[0-5][0-9]:([0-5][0-9]|60)(\.[0-9]+)?"

    def to_python(self) -> float:
        """ """
        self._validate()

        return isodate.parse_time(self)


class FhirCode(FhirPrimitiveType):
    """Indicates that the value is taken from a set of controlled
    strings defined elsewhere (see Using codes for further discussion).
    Technically, a code is restricted to a string which has at least one
    character and no leading or trailing whitespace, and where there is no
    whitespace other than single spaces in the contents

    XML Representation: xs:token
    JSON representation: JSON string
    """

    __visit_name__: str = "code"
    __regex__: str = r"[^\s]+(\s[^\s]+)*"

    def to_python(self) -> str:
        """ """
        self._validate()

        return str(self)


class FhirOid(FhirPrimitiveType):
    """An OID represented as a URI (RFC 3001 ); e.g. urn:oid:1.2.3.4.5

    XML Representation: xs:anyURI
    JSON representation: JSON string - uri
    """

    __visit_name__: str = "oid"
    __regex__: str = r"urn:oid:[0-2](\.(0|[1-9][0-9]*))+"

    def to_python(self) -> str:
        """ """
        self._validate()

        return str(self)


class FhirId(FhirPrimitiveType):
    """Any combination of upper- or lower-case ASCII letters
    ('A'..'Z', and 'a'..'z', numerals ('0'..'9'), '-' and '.', with a length limit
    of 64 characters. (This might be an integer, an un-prefixed OID, UUID or any other
    identifier pattern that meets these constraints.)

    XML Representation: xs:string
    JSON representation: JSON string
    """

    __visit_name__: str = "id"
    __regex__: str = r"[A-Za-z0-9\-\.]{1,64}"

    def to_python(self) -> str:
        """ """
        self._validate()

        return str(self)


class FhirMarkdown(FhirPrimitiveType):
    """A FHIR string (see above) that may contain markdown syntax
    for optional processing by a markdown presentation engine,
    in the GFM extension of CommonMark format (see below)

    About the markdown datatype:
    -   This specification requires and uses the GFM (Github Flavored Markdown)
    extensions on CommonMark  format
    - Note that GFM prohibits Raw HTML
    - Systems are not required to have markdown support, so the content of a
    string should be readable without markdown processing, per markdown philosophy
    - Markdown content SHALL NOT contain Unicode character points below 32, except for
    u0009 (horizontal tab), u0010 (carriage return) and u0013 (line feed)
    - Markdown is a string, and subject to the same rules (e.g. length limit)
    - Converting an element that has the type string to markdown in a later version of
    this FHIR specification is not considered a breaking change (neither is adding
    markdown as a choice to an optional element that already has a choice of data types)

    XML Representation: xs:string
    JSON representation: JSON string
    """

    __visit_name__: str = "markdown"
    __regex__: str = r"\s*(\S|\s)*"

    def to_python(self) -> Text:
        """"""
        return str(self)


class FhirUnsignedInt(FhirPrimitiveType):
    """Any non-negative integer in the range 0..2,147,483,647

    XML Representation: xs:nonNegativeInteger
    JSON representation: JSON number
    """

    __visit_name__: str = "unsignedInt"
    __regex__: str = r"[0]|([1-9][0-9]*)"

    def to_python(self) -> int:
        """ """
        self._validate()

        return int(self)


class FhirPositiveInt(FhirPrimitiveType):
    """Any positive integer in the range 1..2,147,483,647

    XML Representation: xs:positiveInteger
    JSON representation: JSON number
    """

    __visit_name__: str = "positiveInt"
    __regex__: str = r"\+?[1-9][0-9]*"

    def to_python(self) -> int:
        """ """
        self._validate()

        return int(self)


class FhirUUID(FhirPrimitiveType):
    """A UUID (aka GUID) represented as a URI (RFC 4122 );
    e.g. urn:uuid:c757873d-ec9a-4326-a141-556f43239520

    XML Representation: xs:anyURI
    JSON representation: JSON string - uri
    """

    __visit_name__: str = "uuid"
    __regex__: Optional[str] = None

    def to_python(self) -> Text:
        """ """
        return str(self)


# FHIR Primitive Data Types
PrimitiveDataTypes: ImmutableDict = ImmutableDict(
    [
        (FhirBoolean.__visit_name__, FhirBoolean),
        (FhirInteger.__visit_name__, FhirInteger),
        (FhirString.__visit_name__, FhirString),
        (FhirDecimal.__visit_name__, FhirDecimal),
        (FhirURI.__visit_name__, FhirURI),
        (FhirURL.__visit_name__, FhirURL),
        (FhirCanonical.__visit_name__, FhirCanonical),
        (FhirBase64Binary.__visit_name__, FhirBase64Binary),
        (FhirInstant.__visit_name__, FhirInstant),
        (FhirDate.__visit_name__, FhirDate),
        (FhirDateTime.__visit_name__, FhirDateTime),
        (FhirTime.__visit_name__, FhirTime),
        (FhirCode.__visit_name__, FhirCode),
        (FhirOid.__visit_name__, FhirOid),
        (FhirId.__visit_name__, FhirId),
        (FhirMarkdown.__visit_name__, FhirMarkdown),
        (FhirUnsignedInt.__visit_name__, FhirUnsignedInt),
        (FhirPositiveInt.__visit_name__, FhirPositiveInt),
        (FhirUUID.__visit_name__, FhirUUID),
    ]
)


class Empty:
    """Empty Class: specially designed for fhirpath to identify empty value"""

    __slots__ = ()

    def __repr__(self) -> str:
        return "<NO_VALUE>"


EMPTY_VALUE = Empty()


@implementer(IFhirPrimitiveType, IPrimitiveTypeCollection)
class PrimitiveTypeCollection(object):
    """ """

    __visit_name__: str = "collection"
    __regex__: Optional[str] = None
    _container: Deque[FhirPrimitiveType]
    _registered_visit: Optional[str]
    __slots__ = ("_container", "_registered_visit")

    def __init__(self, *members):
        """ """
        object.__setattr__(self, "_container", deque())
        object.__setattr__(self, "_registered_visit", None)

        for member in members:
            self.add(member)

    def add(self, item: FhirPrimitiveType, position: Optional[int] = None):
        """ """
        member = IFhirPrimitiveType(item)
        if self._registered_visit is None:
            self._registered_visit: Optional[str] = member.__visit_name__

        if member.__visit_name__ != self._registered_visit:
            raise ValueError

        if position is None:
            self._container.append(member)
        else:
            self._container.insert(position, member)

    def remove(
        self, item: Optional[FhirPrimitiveType] = None, position: Optional[int] = None
    ):
        """ """
        if item is None:
            assert position is not None, "Position number is required!"
            try:
                item = self._container[position]
            except IndexError:
                raise

        self._container.remove(item)

    @property
    def registered_visit(self) -> Union[str, None]:
        """ """
        return self._registered_visit

    def to_python(self):
        """ """
        return [member.to_python() for member in self._container]

    def __len__(self) -> int:
        """ """
        return len(self._container)

    def __iter__(self):
        """ """
        for member in self._container:
            # validation purpose
            member.to_python()
            yield member


@implementer(ITypeSpecifier)
class TypeSpecifier(str):
    """ """
